server/src/main/java/org/opensearch/tasks/SearchTaskRequestOperationsListener.java

```diff
@@ -0,0 +1,40 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.tasks;
+
+import org.opensearch.action.search.SearchPhaseContext;
+import org.opensearch.action.search.SearchRequestContext;
+import org.opensearch.action.search.SearchRequestOperationsListener;
+
+/**
+ * SearchTaskRequestOperationsListener subscriber for operations on search tasks resource usages.
+ * Listener ensures to refreshResourceStats on request end capturing the search task resource usage
+ * upon request completion.
+ *
+ */
+public final class SearchTaskRequestOperationsListener extends SearchRequestOperationsListener {
+    private static SearchTaskRequestOperationsListener instance;
+    private final TaskResourceTrackingService taskResourceTrackingService;
+
+    private SearchTaskRequestOperationsListener(TaskResourceTrackingService taskResourceTrackingService) {
+        this.taskResourceTrackingService = taskResourceTrackingService;
+    }
+
+    public static synchronized SearchTaskRequestOperationsListener getInstance(TaskResourceTrackingService taskResourceTrackingService) {
+        if (instance == null) {
+            instance = new SearchTaskRequestOperationsListener(taskResourceTrackingService);
```

Comment:
This static instance is weird. It makes the assumption that all calls to this method will pass in the same taskResourceTrackingService instance and will silently do the wrong thing if that assumption ever does not hold in the future. Why is this necessary? Are you sure this is better than the lock-free alternative of creating a new instance on every invocation?

Reply:
@andrross - I had similar thoughts. Please let me know what you think about - https://github.com/opensearch-project/OpenSearch/pull/14832#discussion_r1683547868

Reply:
@andrross - Please let us know your thoughts